### PR TITLE
feat: support python udf

### PR DESCRIPTION
**Pull request by xudong963**
- [x] Make dozer core run python udf
- [x] Add tests
- [x] install virtualenv in ci to run python udf test.

### CODE REVIEW

File: dozer-tests/src/e2e_tests/cases/eth/dozer-config.yaml [hunk 1]
```diff
@@ -12,7 +12,7 @@ connections:
     name: eth_logs
 sql: |
   -- Eth stats table
-  select block_number, sum(id) into eth_stats
+  select block_number, sum(id), py_udf('add', id) into eth_stats
```
Comment (snork-alt) [inline dozer-tests/src/e2e_tests/cases/eth/dozer-config.yaml:15]: Can we have a syntax `py:add()`

File: dozer-tests/src/e2e_tests/cases/eth/dozer-config.yaml [hunk 2]
```diff
@@ -43,3 +43,10 @@ endpoints:
       primary_key:
         - block_number
         - id
+
```
Comment (snork-alt) [inline dozer-tests/src/e2e_tests/cases/eth/dozer-config.yaml:46]: Can we avoid having to define udfs in the config, but rather Discovery them from the .py file itself ?

File: dozer-tests/src/e2e_tests/cases/eth/dozer-config.yaml (hunk 1 above)
Comment (xudong963) [reply dozer-tests/src/e2e_tests/cases/eth/dozer-config.yaml:15]: No problem

File: dozer-tests/src/e2e_tests/cases/eth/dozer-config.yaml (hunk 2 above)
Comment (xudong963) [reply dozer-tests/src/e2e_tests/cases/eth/dozer-config.yaml:46]: If not defined udfs in config, should we require users to put .py file in a specific location?  cc @mediuminvader

Comment (snork-alt) [reply dozer-tests/src/e2e_tests/cases/eth/dozer-config.yaml:46]: Would be nice to have a python decorator

File: dozer-tests/src/e2e_tests/cases/eth/dozer-config.yaml (hunk 1 above)
Comment (xudong963) [reply dozer-tests/src/e2e_tests/cases/eth/dozer-config.yaml]: `:` or `::` will be handled incorrectly by sqlparser and return unintended results. If we still want to use `:`, I'll check sqlparser deeply

**Comment by xudong963** [comment]
Update the status:

Now the ticket works by the following steps:

1. Create virtualenv, such as `virtualenv test`, and activate it: `./test/bin/activate`
2. Create module `py_udf` (currently the module is fixed) in `test`
3. Add some functions in `py_udf`, such as: 
    ```python
   def add(a):
     return a + 1


    def sum(a, b):
      return a + b
    ```
4. Use python udf in dozer, such as `SELECT py_add(a, 'FLOAT'), py_sum(a, b, 'FLOAT') from t1;` (The last arg in function is return type).

Table t1, with a, b field, and values 1, 2, will get the results:
image


Question: how should we test the ticket? I think current sql test framework can't test python udf because sqlite doesn't the sql syntax.

My thought is we can change the sql test framework and make it accept self-defined`source_result`.

**Comment by snork-alt** [comment]
Feel free to modify the test framework to test python execution. Instead of having to specify the return type during the invocation of the function, can we specify as an annotation in the definition of the function itself? Something like:

```
@dozer_type(Int)
def add(a):
 return a + 1
```

**Comment by xudong963** [comment]
> Instead of having to specify the return type during the invocation of the function, can we specify as an annotation in the definition of the function itself? Something like:

Do you mean let users write `@dozer_type(Int)` in their python file function definition ahead? If so, I'd more like the current way.

Imagine in users' python files

```python
@dozer_type(Int)
def ...

@dozer_type(..)
def ...

...

```
This can cause a certain level of intrusion on the user side, and it's not good for us to get this information.

For the current way, the user only needs to know our constraint (the last parameter is the return type), it is easy to write the return type to the sql, and we can easily get this information only during the parse process, the intrusion to our code is almost zero.

File: .github/workflows/coverage.yaml [hunk 3]
```diff
@@ -53,7 +53,10 @@ jobs:
           restore-keys: |
             coverage-${{ runner.os }}-cargo-${{ hashFiles('Cargo.lock') }}
             coverage-${{ runner.os }}-cargo-
-
+      - name: Create python virtual env
+        shell: bash
+        run: |
+          source ./dozer-tests/python_udf/virtualenv.sh
```
Comment (xudong963) [inline .github/workflows/coverage.yaml:59]: Need Help

I want to create virtual env to run python udf test in it, but seems the script doesn't work. 

Do you have some thoughts?

Comment (chubei) [reply .github/workflows/coverage.yaml:59]: Looks fine to me. Don't know why it doesn't work.

Comment (xudong963) [reply .github/workflows/coverage.yaml:59]: Yeah, also works in my local

Comment (mediuminvader) [reply .github/workflows/coverage.yaml:59]: are each action steps executed in the same shell? I mean if the env is activated at "Create python virtual env" step, will be still active at "Run actions-rs/cargo@v1" step?

Comment (chubei) [reply .github/workflows/coverage.yaml:59]: Ah I think @mediuminvader got it here. Each step probablys runs in a different shell.

Comment (xudong963) [reply .github/workflows/coverage.yaml:59]: Let me think how to fix it later :)

Comment (xudong963) [reply .github/workflows/coverage.yaml]: done 😎

**Review by snork-alt — APPROVED** [review]

**Comment by xudong963** [comment]
Let's merge and iterate later

**Pull request merged**